### PR TITLE
UCT/IB: Add ports check during md_open

### PR DESCRIPTION
**Pull request by ivankochin**
## What
Additional ports availability check during `md_open`.

## Why ?
The ports availability should be checked as soon as possible to make customer experience better.

### CODE REVIEW

**Comment by yosefe** [comment]
@ivankochin where is the failing ibv_create_cq() coming from?
also need to align the code style

**Comment by ivankochin** [comment]
From here: https://github.com/openucx/ucx/blob/9365d7dcaca2f4ca5bcbbf5d641c781078892e68/src/uct/ib/mlx5/dv/ib_mlx5_dv.c#L64

File: src/uct/ib/base/ib_device.c [hunk 1]
```diff
@@ -1095,6 +1095,21 @@ ucs_status_t uct_ib_modify_qp(struct ibv_qp *qp, enum ibv_qp_state state)
...
+
+    for (port_num = dev->first_port; port_num < dev->first_port + dev->num_ports;
+         ++port_num)
+    {
+        status = uct_ib_device_port_check(dev, port_num, 0);
+        if (status == UCS_OK) {
+            return UCS_OK;
+        }
+    }
+    return UCS_ERR_UNREACHABLE;
```
Comment (gleon99) [inline src/uct/ib/base/ib_device.c:1110]: Maybe `return status` instead?

File: examples/uct_hello_world.c [hunk 2]
```diff
@@ -346,19 +346,24 @@ static ucs_status_t dev_tl_lookup(const cmd_args_t *cmd_args,
         for (md_index = 0; md_index < component_attr.md_resource_count; ++md_index) {
             status = uct_md_config_read(components[cmpt_index], NULL, NULL,
                                         &md_config);
-            CHKERR_JUMP(UCS_OK != status, "read MD config",
-                        release_component_list);
+            if (UCS_OK != status) {
```
Comment (gleon99) [inline examples/uct_hello_world.c:349]: Are you sure we need to modify this flow and the code after `uct_md_query`?
Isn't it enough to only change the logic for `uct_md_open`?

File: examples/uct_hello_world.c [hunk 3]
```diff
@@ -346,19 +346,24 @@ static ucs_status_t dev_tl_lookup(const cmd_args_t *cmd_args,
...
+                continue;
+            }
 
             status = uct_md_open(components[cmpt_index],
                                  component_attr.md_resources[md_index].md_name,
                                  md_config, &iface_p->md);
             uct_config_release(md_config);
-            CHKERR_JUMP(UCS_OK != status, "open memory domains",
-                        release_component_list);
+            if (UCS_OK != status) {
```
Comment (gleon99) [inline examples/uct_hello_world.c:359]: Seems such changes are needed in several places, see also pipeline (e,g `test_md`).

File: test/gtest/uct/test_md.cc [hunk 4]
```diff
@@ -97,13 +97,20 @@ test_md::test_md()
 
 void test_md::init()
 {
+    uct_md_h h;
```
Comment (gleon99) [inline test/gtest/uct/test_md.cc:100]: `h` -> `md`

File: src/uct/ib/base/ib_device.c (hunk 1 above)
Comment (ivankochin) [inline src/uct/ib/base/ib_device.c:1110]:
```suggestion
    return status;
```

File: examples/uct_hello_world.c [hunk 5]
```diff
@@ -353,12 +353,14 @@ static ucs_status_t dev_tl_lookup(const cmd_args_t *cmd_args,
...
+                fprintf(stderr, "Failed to open MD: %s\n",
+                        component_attr.md_resources[md_index].md_name);
+                continue;
+            }
 
             status = uct_md_query(iface_p->md, &iface_p->md_attr);
-            CHKERR_JUMP(UCS_OK != status, "query iface",
-                        close_md);
+            CHKERR_JUMP(UCS_OK != status, "query iface", close_md);
```
Comment (gleon99) [inline examples/uct_hello_world.c:363]: Minor: maybe fix `query iface` -> `query MD`?

File: src/uct/ib/base/ib_device.c [hunk 6]
```diff
@@ -1095,6 +1095,19 @@ ucs_status_t uct_ib_modify_qp(struct ibv_qp *qp, enum ibv_qp_state state)
     return UCS_OK;
 }
 
+ucs_status_t uct_ib_device_check_ports(uct_ib_device_t *dev) {
+    ucs_status_t status = UCS_ERR_UNREACHABLE;
```
Comment (gleon99) [inline src/uct/ib/base/ib_device.c:1099]: Is the initialization needed?

**Comment by ivankochin** [comment]
It was decided that we cannot fail md_open call in case of absence of the available ports.

**Pull request closed without merging**